Refactor/clarify-input-files (#278)
AlphaGenes/AlphaPeel: docs/source/usage.rst

      -seq_file [SEQ_FILE ...]
                          Sequence allele read count file(s)
                          (:ref:`see format details <seq_file_format>`).
      -hap_file [HAP_FILE ...]
                          Haplotype file 
                          (:ref:`see format details <hap_file_format>`)
      -x_chr              Indicate that input data is for the :ref:`X chromosome <zero_one_two_etc>`.
      -pheno_file [PHENO_FILE ...]
                          Phenotype file(s)
                          (:ref:`see format details <pheno_file_format>`).
      -plink_file [PLINK_FILE ...]
                          Plink (binary) file(s)
                          (:ref:`see format details <plink_file_format>`).
      -phased_geno_prob_file [PHASED_GENO_PROB_FILE ...]
                          Optional external phased_geno_prob_file file(s) 
                          (:ref:`see format details <penetrance_file_format>`).

>                           This will overwrite the default genotype probability values. 

Reviewer: ```suggestion
                          This will provide the starting internal genotype probability state. 
```

Would this work better in line with our discussion?

      -pheno_file [PHENO_FILE ...]
                          Phenotype file(s)
                          (:ref:`see format details <pheno_file_format>`).

>       -hap_file [HAP_FILE ...]

Reviewer: Since this is a genetic information file I think it best if we put it before the -pheno_file, but best to add it after the -seq_file since -hap_file is a derived file, not primary data.

                          Sequence error :ref:`probability <prob_freq_rate>`.
                          Must not be 0.
                          Default: 0.001.

>       -penetrance [PENETRANCE ...]

Reviewer: Aha, this is the "hidden" non-documented option - thanks for brining it up into docs @XingerTang! We should be clearer about the name in line with the other terms we have used for the other input files - would you do this as part to this PR or follow up? What arg name would you pick?

Author: I was also thinking about it, ``penetrance_file`` or ``geno_prob_file`` could be a good choice.

Reviewer: I think geno_prob_file would indeed be the best. Remind me please, are these “penetrances” phased or non-phased? I think they are phased, but could be wrong. I am asking as we have two geno prob outputs and it would be good to sync this arg name with that: geno_prob and phased_geno_prob. So should this arg be geno_prob_file or phased_geno_prob_file?

Author: These are phased, yes, it would be better if we use ``phased_geno_prob_file``.

Reviewer: Looking at all the options, including the outputs, where we have `-phased_geno_prob` to write out phased geno probs into a file, then indeed we should call this input file option as `phased_geno_prob_file`.

                          Optional external :ref:`penetrance <penetrance_file_format>` file(s). 

>                           This will overwrite the default penetrance values.

Reviewer: Do we need to say more as we have individual penetrance probs and model parameters so it gets confusing ...

Author: I put it here just because the options groups in the code are decided so. But I think it should be fine if I move this option to the group of other input file options.

Reviewer: If this is input for individuals then best to move it up with other inputs for individuals. We should be very clear about the overwriting in a clear way (here or elsewhere).

|Software| supports the following genomic data files:
genotype file in the AlphaGenes format (``-geno_file``),
sequence allele read count file in the AlphaGenes format (``-seq_file``),
haplotype file in the AlphaGenes format (``-hap_file``), and
binary Plink file (``-plink_file``).

> |Software| supports genotype probabilities input in the AlphaGenes format (``-penetrance``). 

Reviewer: Sync arg name and there is “phased?” that needs polish

> The ``-penetrance`` input assumes the genotype probability data is error-free. 

Reviewer: Are these individual geno probs stored in penetrance probs? I think so, but check. If yes, then we should probably write something like `The ``-penetrance`` input is assumed to contain error-free phased? genotype probabilities. Hence, you should not ...`

> Hence, when using ``-penetrance`` as the input, you should not use with options ``-est_geno_error_prob``, 

Reviewer: @XingerTang I am rereading these changes - so, we input phased geno probs, which gives us an initial state. I could as well input observed genotype data and when we propagate that observed information the phased geno probs will change from the initial state. I am not looking at the code, just trying to understand the logic.

If the above is correct, why shouldn't I estimate geno_error_prob or any other model parameters? In principle I could, but we should likely caution users that this might be unwise so they avoid double counting the information (if phased geno probs came from the same data) - I assume this is what you wanted to warn here!? Can you rephrase to this effect or lets clarify what the real "issue" for this "warning note" is.

Author: I didn't write the warnings myself, it is there from previous developers. 
I guess the reason why we can't do the estimation with input phased geno prob is that the errors are estimated by the difference between genotype probabilities and genotypes; we are not guaranteed to have genotypes input if we are using genotype probabilities as the input. 
Even if we have the genotype input, the genotype error is used in a way that updates/rewrites the genotype probabilities based on the genotypes and updated genotype errors per peeling cycle. The penetrance input would be fully overwritten after a single cycle.

Reviewer: > I didn't write the warnings myself, it is there from previous developers.
Noted.

> I guess the reason why we can't do the estimation with input phased geno prob is that the errors are estimated by the difference between genotype probabilities and genotypes; we are not guaranteed to have genotypes input if we are using genotype probabilities as the input.
Hmm, but the same applies when we use some observed genotype data just for some individuals, so I think this is not a blocker, in principle. I guess the question is what happens when we don't provide genotype data at all!? Can you check what happens in that case (with -est_geno_error...) - I assume that should fail or produce strange result!?

> Even if we have the genotype input, the genotype error is used in a way that updates/rewrites the genotype probabilities based on the genotypes and updated genotype errors per peeling cycle. The penetrance input would be fully overwritten after a single cycle.
Yes, this is what I had thought - we usually read in genotypes and calculate genotype probs based on that and pedigree. Here with genotype probs as input, we have the genotype probs immediately, however if we would bring in some additional genotype data (say the new batch), then by propagating that new data up the pedigree and then the updated genotype probs down the pedigree, we would get new/updated genotype probabilities. So, this makes sense, in principle, right?

I do agree we need some sort of a warning, the question just is how to word it. Maybe we write in 2-4 sentences what this import does and what happens so people will understand the implications! I will work with you on this, just draft something and we can iterate.

Author: AlphaPeel would automatically set the parameter estimation flags to false if the phased genotype probability is provided, with an error message. 

In ``PeelingInfo.py``:
```
    if args.phased_geno_prob_file is not None:
        if peelingInfo.isXChr:
            warnings.warn(
                "Using an external phased genotype probability file and the x_chr option is highly discouraged. Please do not use."
            )

        if args.est_geno_error_prob:
            warnings.warn(
                "External phased genotype probability file included, but est_geno_error_prob flag used. The two options are incompatible. est_geno_error_prob set to false."
            )
            args.est_geno_error_prob = False

        if args.est_seq_error_prob:
            warnings.warn(
                "External phased genotype probability file included, but est_seq_error_prob flag used. The two options are incompatible. est_seq_error_prob set to false."
            )
            args.est_seq_error_prob = False
```

Reviewer: Ok, I see now. Let’s keep the behaviour and adapt it if and as needed in the future.

|Software| supports the following genomic data files:
genotype file in the AlphaGenes format (``-geno_file``),
sequence allele read count file in the AlphaGenes format (``-seq_file``),
haplotype file in the AlphaGenes format (``-hap_file``), and
binary Plink file (``-plink_file``).

.. note::

  Internally, the haplotype input by ``-hap_file`` is used to fill up the missing genotype information,
  and the phasing information is not preserved. If you want to input phased data, 
  consider the ``-phased_geno_prob_file`` input.

|Software| supports genotype probabilities input in the AlphaGenes format (``-phased_geno_prob_file``). 
The ``-phased_geno_prob_file`` is assumed to contain error-free phased genotype probabilities. 
Hence, when using ``-phased_geno_prob_file`` as the input, you should not use with options ``-est_geno_error_prob``, 
``-est_seq_error_prob``. 
The option does not support the ``-x_chr`` option for now. 

When ``-x_chr`` is used the :ref:`pedigree file <ped_file_format>` and
:ref:`genotype file <geno_file_format>` have specific requirements.

> Follow the links to e file formats for more details.

Reviewer: ```suggestion
Follow the links to file formats for more details.
```

                            Value(s) less than 1/3 are replaced by 1/3.
      -geno_prob            Output :ref:`genotype probabilities <geno_prob_file_format>`.
      -phased_geno_prob     Output :ref:`phased genotype probabilities <phased_geno_prob_file_format>`.

>       -hap                  Call and output :ref:`haplotypes <haplotype_file_format>`.

Reviewer: I think you want to revert this back since the tag has the short form and I think we use short form to the tags.

Reviewer: Aha, there are two, ugh, tricky. Let me check this out.

Reviewer: OK, I see you followed the previous format with geno/genotype_file_format, which is good.

  id4 2 0 6 7
  id4 0 7 7 0

.. _hap_file_format:

Haplotype file
==============

This file has two lines with

> *observed phased haplotypes* for each phased genotyped individual; 

Reviewer: ```suggestion
*observed haplotypes* for each phased genotyped individual; 
```

possibly from previous analyses.

> The sequence of the maternal and paternal haplotypes won't affect the peeling.

Reviewer: ```suggestion
The order of maternal and paternal haplotypes is not important
because the current implementation translates these haplotypes into genotypes.
```

  id3 1
  id4 0

> .. _hap_file_format:

Reviewer: Move above pheno file

Haplotype file
==============

This file has two lines with

> *observed phased haplotypes* for each phased genotyped individual.

Reviewer: ```suggestion
This file has two lines with
*observed phased haplotypes* for each phased genotyped individual;
possibly as an external reference haplotype library or
from previous analyses.
```

@XingerTang what are your thoughts on this suggestion?

Author: This is safer to do than ``penetrance`` as it will consider the other inputs alongside. But I don't think it is designed to be used as the reference haplotype library.

Reviewer: Thinking a bit more, this should likely be:

```
This file has two lines with
*observed haplotypes* for each genotyped and phased individual;
possibly from previous analyses.
```

@XingerTang let's talk next week how this input is actually used - can you point me in the code where it is used?

Author: The ``hap_file`` is read into the pedigree by ``tinyhouse.Pedigree.Pedigree.readInPhase``, the information would be stored in the ``haplotypes`` attribute of the ``Pedigree`` object before the peeling. 

Later, at line 88 of ``PeelingInfo.py``, 
```
HaplotypeOperations.ind_fillInGenotypesFromPhase(ind)
```
It will convert the haplotypes to genotypes (ignoring phasing information) and calculate genotype probabilities based solely on the genotypes.

Reviewer: Ah, ok, so this is then not very powerful. OK, I propose the following.

1) Write somewhere what happens with these haplotypes - that they are imported but eventually combined and used as genotypes. Not powerful (since phase information is lost), but a capability if anyone wants it.

2) Open an issue to develop a function that calculates phased genotype probabilities from the haplotypes - this should not be hard and will not throw away the phase information. Do you agree?

The first line is for paternal haplotype, and the second line is for maternal haplotype.
The file does not need to include all individuals present in other files.
The first value in each line is the individual's ID.
The remaining values are observed haplotypes at each locus,
encoded as ``0`` and ``1``, or ``9`` when the value is missing. 
Only loci on one chromosome should be provided!

Example with four individuals and four loci:

::

  id1 0 1 1 0 # Paternal haplotype
  id1 0 1 1 0 # Maternal haplotype
  id2 1 0 1 0
  id2 0 1 0 1
  id3 0 1 1 0
  id3 1 0 1 0
  id4 0 1 1 0
  id4 0 1 0 0

> The X chromosome functionality for haplotype input is not tested. 

Reviewer: ```suggestion
The X chromosome functionality for haplotype input is not tested, so do not use it. 
```

Reviewer: Best to open issue about this so we track it.

Author: Issue is opened: https://github.com/AlphaGenes/AlphaPeel/issues/279

.. _penetrance_file_format:

Penetrance file
===============

> This file has four lines with *phased genotype probabilities* for each individual.

Reviewer: ```suggestion
This file has four lines with
*phased genotype probabilities* for each individual;
possibly as an external probabilistic reference haplotype library or
from previous analyses.
```

@XingerTang would this work / be correct?

Author: ``-penetrance`` input assumes ``0`` error and goes directly to the genotype probabilities, while overwriting other inputs if there is any overlap in the data. I think it would be quite risky to use the reference haplotype library here.

Reviewer: Taken together this should likely be:

```
Phased genotype probabilities file
==================================

This file has four lines with
*phased genotype probabilities* for each genotyped and phased individual;
possibly from previous analyses.
The four lines respectively correspond to probabilities for
``aa``, ``aA``, ``Aa``, and ``AA`` phased genotypes.
The file does not need to include all individuals present in other files.
The first value in each line is the individual ID.
The remaining values are phased genotype probabilities at each locus.
These values are assumed without an error and
used directly as individual's phased genotype probability state.
```

> The four lines correspond to probabilities for

Reviewer: The example below has 3 lines per ind;)
